tests/Optimizer/Transformer/AutoExtensionsTest.php

```diff
@@ -0,0 +1,193 @@
+                . '    <script async custom-element="amp-access" src="https://cdn.ampproject.org/v0/amp-access-0.1.js"></script>'
+                . '    <script async custom-element="amp-analytics" src="https://cdn.ampproject.org/v0/amp-analytics-0.1.js"></script>'
+                . '    <script async custom-element="amp-access-scroll" src="https://cdn.ampproject.org/v0/amp-access-scroll-0.1.js"></script>'
+                . '    <script id="amp-access" type="application/json">'
+                . '        ['
+                . '            {'
+                . '                "vendor": "scroll",'
+                . '                "namespace": "scroll"'
+                . '            }'
+                .          ']'
+                . '    </script>'
+                . '</head><body>'
+                . '    <section amp-access="NOT scroll.scroll">'
+                . '        ad goes here'
+                . '    </section>'
+                . '</body></html>',
+            ],
+
+            'keep unneeded amp-youtube extension when removeUnneededExtensions is false' => [
+                TestMarkup::DOCTYPE . '<html><head>' . TestMarkup::META_CHARSET
+                . '<script async custom-element="amp-youtube" src="https://cdn.ampproject.org/v0/amp-youtube-0.1.js"></script>'
+                . '<script async custom-element="amp-anim" src="https://cdn.ampproject.org/v0/amp-anim-0.1.js"></script>'
+                . '</head><body><amp-anim></amp-anim></body></html>',
+                TestMarkup::DOCTYPE . '<html><head>' . TestMarkup::META_CHARSET
+                . '<script async custom-element="amp-youtube" src="https://cdn.ampproject.org/v0/amp-youtube-0.1.js"></script>'
+                . '<script async custom-element="amp-anim" src="https://cdn.ampproject.org/v0/amp-anim-0.1.js"></script>'
+                . '</head><body><amp-anim></amp-anim></body></html>',
+            ],
+
+            'remove unneeded amp-youtube extension when removeUnneededExtensions is true' => [
+                TestMarkup::DOCTYPE . '<html><head>' . TestMarkup::META_CHARSET
+                . '<script async custom-element="amp-youtube" src="https://cdn.ampproject.org/v0/amp-youtube-0.1.js"></script>'
+                . '<script async custom-element="amp-anim" src="https://cdn.ampproject.org/v0/amp-anim-0.1.js"></script>'
+                . '</head><body><amp-anim></amp-anim></body></html>',
+                TestMarkup::DOCTYPE . '<html><head>' . TestMarkup::META_CHARSET
+                . '<script async custom-element="amp-anim" src="https://cdn.ampproject.org/v0/amp-anim-0.1.js"></script>'
+                . '</head><body><amp-anim></amp-anim></body></html>',
+                [],
+                [
+                    AutoExtensionsConfiguration::REMOVE_UNNEEDED_EXTENSIONS => true,
+                ]
+            ],
+
+            'keep unneeded amp-carousel extension when removeUnneededExtensions is true' => [
```

Comment:
`amp-carousel` should only be kept if `amp-lightbox-gallery` is present (even if no `amp-carousel` tag is found).